### ISSUE TITLE
Is JournalInitialPosition.START_FROM_CURRENT works properly for clustered instances?

### ISSUE THREAD

**Issue by Ultar** [hazelcast/hazelcast-jet#2420]
Hello!

I'm sourcing my stream from map journal. When the map journal is not empty, and I'm restarting my jobs in single Jet instance, it works properly with START_FROM_CURRENT and START_FROM_OLDEST initial values. But when the START_FROM_CURRENT is set and I'm turn off or restart one node in cluster, then my streams performs jobs processing once again, as if the START_FROM_OLDEST is enabled. The docs says that jobs will restart automatically when scaling is performed. What I'm doing wrong?

I use the hazelcast-jet-spring:3.2.2 and currently have no possibility to migrate into version 4.

Thanks!

**Comment by viliam-durina** [comment]
You don't say whether you have a processing guarantee enabled, but it should be the other way:
- if processing guarantee is enabled, the first execution starts according the initial position option, after a restart we always start from the snapshotted offset
- if processing guarantee is not enabled, every execution starts from the intial offset (either the oldest or currnet).

I've tested the behavior, it works as I described. Could you please provide a test case where you ask to START_FROM_CURRENT, but after a restart it starts from oldest?

**Comment by viliam-durina** [comment]
This is the test I did:
```java
        Pipeline p = Pipeline.create();
        p.readFrom(Sources.mapJournal(map, START_FROM_OLDEST))
         .withoutTimestamps()
         .writeTo(Sinks.logger());

        Job job = instance.newJob(p);
        assertJobStatusEventually(job, JobStatus.RUNNING);
        logger.info("job running");
        // wait for a while and add some items to the journal
        sleepSeconds(1);
        map.put("1", 1);
        map.put("2", 2);

        // give time so that the job can print them
        sleepSeconds(1);

        // now restart. If initial=OLDEST, the two events should be printed again. 
        // If initial=CURRENT, they should be not printed again.
        logger.info("restarting...");
        job.restart();

        sleepSeconds(1);
        
        job.cancel();
```

**Comment by Ultar** [comment]
Yes, your test works as expected for me too, but I meant that problem happens in clustered environment (we have 2 HZ Jet instances and we were restarting each of them). Also we enabled ProcessingGuarantee.EXACTLY_ONCE and this has disabled the additional execution of jobs.

This is our pipeline:
`StreamStage<BusinessRulesDecision> updatedDecisionObjects =
            p.drawFrom(Sources.mapJournal(MapNames.IMAP_BUSINESS_RULES_DECISION_TABLE,
                Util.mapPutEvents(),
                Util.mapEventNewValue(),
                JournalInitialPosition.START_FROM_CURRENT))
                .withoutTimestamps()
                .map(entry -> (BusinessRulesDecision) entry);

...

        actionEventStream
            .filter(a -> ActionTypes.HARDWARE_COMMAND.getId() == a.getRefType())
            .mapUsingIMap(MapNames.IMAP_BUSINESS_RULES_ACTIONS_HARDWARE_COMMANDS_CACHE,
                BusinessRuleActionEvent::getTenantId,
                BusinessRuleActionProcessor::transformIntoHardwareCommand)
            .drainTo(BusinessRuleActionsSinks.newHardwareCommand());

public class BusinessRuleActionsSinks {
    public static Sink<Object> newHardwareCommand() {
        return SinkBuilder.sinkBuilder("actions-hardware-command-sink",
            context -> context.jetInstance().getHazelcastInstance().getTopic(Topics.LISTENER_HARDWARE_COMMAND))
            .receiveFn(ITopic::publish)
            .build();
    }`

**Comment by viliam-durina** [comment]
What do you mean by this? 
>we enabled ProcessingGuarantee.EXACTLY_ONCE and this has disabled the additional execution of jobs.

**Comment by Ultar** [comment]
> What do you mean by this?
> 
> > we enabled ProcessingGuarantee.EXACTLY_ONCE and this has disabled the additional execution of jobs.

I meant that setting of this resolved the issue. But I don't see the connection to source problem.

**Comment by viliam-durina** [comment]
I tested again with no processing guarantee and it works as expected. This time I used 2 members, started the job on them. Then i shut down one of them and later I started another member so that the cluster again had 2 members. It worked as expected: in case of OLDEST the same events were processed after each job restart, in case of CURRENT the events were logged only in the first execution.

However I tested on current `master` branch. Then I also tried the 3.2.2 release, the behavior was the same.

**Comment by Ultar** [comment]
Thanks @viliam-durina for your tests!

I found this in javadoc for no processing guarantee:

`     * No processing guarantees are given and no snapshots are taken during
     * job execution. When a job is restarted automatically it will be as
     * if the job is starting from scratch which can cause items to be lost
     * or duplicated.`

So, probably it was our issue when items were duplicated. We also checked metrics which are responsible for emmiting events from map journals and there were no spikes of events.
Anyway the EXACTLY_ONCE setting helped us and thank you again!
